### PR TITLE
pass around chainhash.Hash versus strings

### PR DESCRIPTION
**Pull request by dajohi**

### CODE REVIEW

File: mempool/collector.go [hunk 1]
```diff
@@ -112,7 +118,7 @@ func (t *MempoolDataCollector) mempoolTxns() ([]exptypes.MempoolTx, error) {
 			VoutCount: len(msgTx.TxOut),
 			Vin:       exptypes.MsgTxMempoolInputs(msgTx),
 			Coinbase:  blockchain.IsCoinBaseTx(msgTx),
-			Hash:      hash,
+			Hash:      hash.String(),
```
Comment (chappjc) [inline mempool/collector.go]: We already have `hashStr`.  Why make a chainhash.Hash in this function?

Comment (dajohi) [reply mempool/collector.go]: reverted to hashStr

**Comment by dajohi** [comment]
I figured since we now generate a Hash with chainhash.NewHashFromStr on user input, we just pass that around versus a string.  It does shuffle where conversions happen, but I feel a bit safer.

File: db/dcrsqlite/apisource.go [hunk 2]
```diff
@@ -1815,7 +1799,7 @@ func (db *WiredDB) GetMempool() []exptypes.MempoolTx {
 		}
 		txs = append(txs, exptypes.MempoolTx{
 			TxID:     msgTx.TxHash().String(),
-			Hash:     hash,
+			Hash:     hash.String(),
```
Comment (chappjc) [inline db/dcrsqlite/apisource.go]: `hashStr` here too pls.

File: middleware/apimiddleware.go [hunk 3]
```diff
@@ -137,32 +137,37 @@ func GetRawHexTx(r *http.Request) string {
...
 	}
 	return hash, nil
 }
 
 // GetTxnsCtx retrieves the ctxTxns data from the request context. If not set,
 // the return value is an empty string slice.
-func GetTxnsCtx(r *http.Request) ([]string, error) {
-	hashes, ok := r.Context().Value(ctxTxns).([]string)
-	if !ok || len(hashes) == 0 {
+func GetTxnsCtx(r *http.Request) ([]*chainhash.Hash, error) {
```
Comment (chappjc) [inline middleware/apimiddleware.go]: I have to admit I don't have an good idea about timing of `NewHashFromStr`, but it should be lightning quick.  Hopefully doesn't add more than 1ms of latency to the middleware.

File: db/dcrsqlite/apisource.go (hunk 2 above)
Comment (dajohi) [reply db/dcrsqlite/apisource.go]: changed to hashStr

**Comment by chappjc** [comment]
In lieu of proper middleware tests and benchmarks, here are the `apiload` results for the "rotation" attack, which hits the regular API at 200 req/sec.

`master`:

```
$   ./apiload -a rotation --frequency 200
2019/03/09 17:13:42 Beginning rotation. Duration: 60 seconds.
Requests      [total, rate]            12000, 200.02
Duration      [total, attack, wait]    1m0.004106759s, 59.99519479s, 8.911969ms
Latencies     [mean, 50, 95, 99, max]  15.352379ms, 808.228µs, 94.728865ms, 111.159092ms, 127.791981ms
Bytes In      [total, mean]            1037856, 86.49
Bytes Out     [total, mean]            0, 0.00
Success       [ratio]                  100.00%
Status Codes  [code:count]             200:12000  
Error Set:
```

This PR rebased on `master`:

```
$   ./apiload -a rotation --frequency 200
2019/03/09 17:15:52 Beginning rotation. Duration: 60 seconds.
Requests      [total, rate]            12000, 200.02
Duration      [total, attack, wait]    1m0.004407305s, 59.995098162s, 9.309143ms
Latencies     [mean, 50, 95, 99, max]  15.520305ms, 813.157µs, 95.394494ms, 111.669316ms, 230.387732ms
Bytes In      [total, mean]            1037768, 86.48
Bytes Out     [total, mean]            0, 0.00
Success       [ratio]                  100.00%
Status Codes  [code:count]             200:12000  
Error Set:
```

**Pull request merged**